proposals/0083-relax-entry-constraints.md

```diff
@@ -0,0 +1,62 @@
+---
+simd: '0083'
+title: Relax Entry Constraints
+authors:
+  - Andrew Fitzgerald (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-11-02
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Remove the constraint that transaction entries must not conflict with each
+other.
+
+## Motivation
+
+The current constraint that transaction entries must not conflict with each
+other is unnecessary.
+It is a constraint, that made the current labs-client implementation easier,
+but is not required for the protocol to function correctly.
+The removal of this constraint simplifies the protocol, and allows more
+flexibility in the implementation of both block-produciton and
+block-validation.
+
+## Alternatives Considered
+
+1. Do nothing
+    - This is the simplest option, as we could leave the protocol as is.
+    However, this leaves the protocol more complex than it needs to be.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+Currently, if a transaction entry within a block contains transactions that
+conflict, the entire block is marked invalid.
+The proposal is that this constraint is removed entirely, and entries are
+allowed to contain transactions that conflict with each other.
+
+## Impact
+
+- The protocol is simplified
+- Block-production is more flexible
+
+## Security Considerations
+
+None
+
+## Drawbacks
+
+None
```

Comment:
second that

Reply:
@Huzaifa696 - Per the below snippet, a self-conflicting entry is currently considered invalid, and will raise an error:
https://github.com/solana-labs/solana/blob/230779d4590ac625e6710117987ef505c0cf71c6/ledger/src/blockstore_processor.rs#L606-L619

That error results in the slot getting marked dead by `ReplayStage`:
https://github.com/solana-labs/solana/blob/230779d4590ac625e6710117987ef505c0cf71c6/core/src/replay_stage.rs#L2782-L2786

If the "rules" are changed to allow a self-conflicting entry, then a slot with a self-conflicting entry will no longer error / result in dead slot.

Suppose that the change to allow self-conflicting entries first lands in `v1.18`. Nodes running `v1.17` will mark slots with a self-conflicting entry dead whereas nodes running `v1.18` will allow it. This is obviously problematic during an upgrade period where the stake is evenly distributed between `v1.17` and `v1.18`. As such, the change to allow self-conflicting entries would have to sit behind a feature gate, and only get enabled once the entire cluster has adopted `v1.18`.

Reply:
> So, as far as I understand, the flow is: approved SIMD -> feature gate issue -> PR (that implements the SIMD with a feature gate). Is that right?

I believe so, yes

Reply:
> So, as far as I understand, the flow is: approved SIMD -> feature gate issue -> PR (that implements the SIMD with a feature gate). Is that right? @steviez

@jacobcreech would know best.

> @lheeger-jump can you please elaborate how will it break the consensus.

As @steviez mentioned, this changes the validation criteria for blocks, which affects whether they are rejected. It needs a feature gate which implies coordination with other clients.

Reply:
> approved SIMD -> feature gate issue -> PR

That would be ideal to avoid any work that may be thrown away due to never getting consensus on a SIMD.